File: ios/RoutineChart/RoutineChart/Core/Data/Remote/Firebase/SyncCursorManager.swift
//
//  SyncCursorManager.swift
//  RoutineChart
//
//  Created for Phase 3.1: Sync Infrastructure
//  Manages sync cursors for tracking last sync timestamp per collection
//

import Foundation
import GRDB
import OSLog

/// Manages sync cursors for tracking last sync timestamp per collection
final class SyncCursorManager {
    private let dbManager: SQLiteManager
    
    init(dbManager: SQLiteManager = .shared) {
        self.dbManager = dbManager
    }
    
    /// Get the last sync timestamp for a collection
    /// Returns nil if no cursor exists (first sync)
    func getCursor(collection: String) async throws -> SyncCursor? {
        let db = try dbManager.database()
        return try await db.read { db in
            try SyncCursor.fetchOne(db, key: collection)
        }
    }
    
    /// Update or create a cursor for a collection
    func updateCursor(collection: String, lastSyncedAt: Date) async throws {
        let db = try dbManager.database()
        try await db.write { db in
            var cursor = SyncCursor(collection: collection, lastSyncedAt: lastSyncedAt)
            try cursor.save(db)
        }
        AppLogger.database.info("Updated sync cursor for collection '\(collection)' to \(lastSyncedAt)")
    }
    
    /// Delete a cursor (useful for resetting sync state)
    func deleteCursor(collection: String) async throws {
        let db = try dbManager.database()
        try await db.write { db in
            try SyncCursor.deleteOne(db, key: collection)
        }
        AppLogger.database.info("Deleted sync cursor for collection '\(collection)'")
    }
    
    /// Get all cursors (for debugging/monitoring)
    func getAllCursors() async throws -> [SyncCursor] {
        let db = try dbManager.database()
        return try await db.read { db in
            try SyncCursor.fetchAll(db)
        }
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/data/remote/firebase/SyncCursorManager.kt
package com.HammersTech.RoutineChart.core.data.remote.firebase

import com.HammersTech.RoutineChart.core.data.local.room.dao.SyncCursorDao
import com.HammersTech.RoutineChart.core.data.local.room.entities.SyncCursorEntity
import com.HammersTech.RoutineChart.core.domain.models.SyncCursor
import com.HammersTech.RoutineChart.core.utils.AppLogger
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages sync cursors for tracking last sync timestamp per collection
 * Phase 3.1: Sync Infrastructure
 */
@Singleton
class SyncCursorManager
    @Inject
    constructor(
        private val syncCursorDao: SyncCursorDao,
    ) {
        /**
         * Get the last sync timestamp for a collection
         * Returns null if no cursor exists (first sync)
         */
        suspend fun getCursor(collection: String): SyncCursor? {
            val entity = syncCursorDao.getByCollection(collection)
            AppLogger.Database.info("Getting cursor for collection '$collection': ${if (entity != null) "found" else "not found"}")
            return entity?.toDomain()
        }

        /**
         * Update or create a cursor for a collection
         */
        suspend fun updateCursor(
            collection: String,
            lastSyncedAt: Instant,
        ) {
            val entity = SyncCursorEntity.fromDomain(collection, lastSyncedAt)
            syncCursorDao.insertOrUpdate(entity)
            AppLogger.Database.info("Updated sync cursor for collection '$collection' to $lastSyncedAt")
        }

        /**
         * Delete a cursor (useful for resetting sync state)
         */
        suspend fun deleteCursor(collection: String) {
            syncCursorDao.deleteByCollection(collection)
            AppLogger.Database.info("Deleted sync cursor for collection '$collection'")
        }

        /**
         * Get all cursors (for debugging/monitoring)
         */
        suspend fun getAllCursors(): List<SyncCursor> {
            val entities = syncCursorDao.getAll()
            AppLogger.Database.info("Getting all cursors: found ${entities.size} cursor(s)")
            entities.forEach { entity ->
                AppLogger.Database.info("  - ${entity.collection}: ${entity.lastSyncedAt}")
            }
            return entities.map { it.toDomain() }
        }
    }
